core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -667,4 +522,163 @@ public String toString() {
+        store.incRef();
+        try {
+            Future[] runners = asyncSendFiles(store, files, outputStreamFactory);
+            IOException corruptedEngine = null;
+            final List<Throwable> exceptions = new ArrayList<>();
+            for (int i = 0; i < runners.length; i++) {
+                StoreFileMetaData md = files[i];
+                try {
+                    runners[i].get();
+                } catch (ExecutionException t) {
+                    corruptedEngine = handleExecutionException(store, corruptedEngine, exceptions, md, t.getCause());
+                } catch (InterruptedException t) {
+                    corruptedEngine = handleExecutionException(store, corruptedEngine, exceptions, md, t);
+                }
+            }
+            if (corruptedEngine != null) {
+                failEngine(corruptedEngine);
+                throw corruptedEngine;
+            } else {
+                ExceptionsHelper.rethrowAndSuppress(exceptions);
+            }
+        } finally {
+            store.decRef();
+        }
+    }
+
+    private IOException handleExecutionException(Store store, IOException corruptedEngine, List<Throwable> exceptions, StoreFileMetaData md, Throwable t) {
+        logger.debug("Failed to transfer file [" + md + "] on recovery");
+        final IOException corruptIndexException;
+        final boolean checkIntegrity = corruptedEngine == null;
+        if ((corruptIndexException = ExceptionsHelper.unwrapCorruption(t)) != null) {
+            if (checkIntegrity && store.checkIntegrityNoException(md) == false) { // we are corrupted on the primary -- fail!
+                logger.warn("{} Corrupted file detected {} checksum mismatch", shardId, md);
+                corruptedEngine = corruptIndexException;
+            } else { // corruption has happened on the way to replica
+                RemoteTransportException exception = new RemoteTransportException("File corruption occurred on recovery but checksums are ok", null);
+                exception.addSuppressed(t);
+                if (checkIntegrity) {
+                    logger.warn("{} Remote file corruption on node {}, recovering {}. local checksum OK",
+                            corruptIndexException, shardId, request.targetNode(), md);
+                } else {
+                    logger.warn("{} Remote file corruption on node {}, recovering {}. local checksum are skipped",
+                            corruptIndexException, shardId, request.targetNode(), md);
+                }
+                exceptions.add(exception);
+
+            }
+        } else {
+            exceptions.add(t);
+        }
+        return corruptedEngine;
+    }
+
+    protected void failEngine(IOException cause) {
+        shard.engine().failEngine("recovery", cause);
+    }
+
+    Future<Void>[] asyncSendFiles(Store store, StoreFileMetaData[] files, Function<StoreFileMetaData, OutputStream> outputStreamFactory) {
```

Comment:
Another question, what made you decide to go with `Future<Void>` and catch the exception on `.get()` instead of `Future<Throwable>` and catch the exception here and return it?